### PR TITLE
Judgedaemon sleep increase

### PR DESCRIPTION
**Pull request by mvr320**
Based on suggestion by @meisterT

### CODE REVIEW

File: judge/judgedaemon.main.php [hunk 1]
```diff
@@ -220,7 +220,10 @@ function rest_encode_file(string $file, $sizelimit = true) : string
     return base64_encode(dj_file_get_contents($file, $maxsize));
 }
 
-$waittime = 5;
+// Both constants in miliseconds
```
Comment (meisterT) [inline judge/judgedaemon.main.php:223]: nit: milliseconds

File: judge/judgedaemon.main.php [hunk 2]
```diff
@@ -220,7 +220,10 @@ function rest_encode_file(string $file, $sizelimit = true) : string
...
+const INITIAL_WAITTIME =  100;
```
Comment (meisterT) [inline judge/judgedaemon.main.php:224]: consider adding the suffix `_MS` to the variable name to make it easier to see the unit

File: judge/judgedaemon.main.php [hunk 3]
```diff
@@ -576,12 +579,15 @@ function fetch_executable(
...
             $dosleep = false;
+            $waittime = INITIAL_WAITTIME;
             break;
         }
     }
     // Sleep only if everything is "waiting" and only if we're looking at the first endpoint again
     if ($dosleep && $currentEndpoint==0) {
-        sleep($waittime);
+        usleep($waittime);
+        $waittime = $waittime*2>MAXIMAL_WAITTIME ? MAXIMAL_WAITTIME : $waittime*2;
```
Comment (meisterT) [inline judge/judgedaemon.main.php:589]: I would rewrite this as:
```
$waittime = min(MAXIMAL_WAITTIME, $waittime*2);
```

File: judge/judgedaemon.main.php [hunk 4]
```diff
@@ -576,12 +579,15 @@ function fetch_executable(
...
+        logmsg(LOG_NOTICE, "Waittime now: ".$waittime);
```
Comment (meisterT) [inline judge/judgedaemon.main.php:590]: Remove debug message?

File: judge/judgedaemon.main.php [hunk 5]
```diff
@@ -220,7 +220,10 @@ function rest_encode_file(string $file, $sizelimit = true) : string
     return base64_encode(dj_file_get_contents($file, $maxsize));
 }
 
-$waittime = 5;
+// Both constants in milliseconds
+const INITIAL_WAITTIME_MS =  100;
+const MAXIMAL_WAITTIME_MS = 5000;
+$waittime = INITIAL_WAITTIME;
```
Comment (thijskh) [inline judge/judgedaemon.main.php:226]: Variable is not defined

File: judge/judgedaemon.main.php [hunk 6]
```diff
@@ -576,12 +579,14 @@ function fetch_executable(
         }
         if (!$endpoint['waiting']) {
             $dosleep = false;
+            $waittime = INITIAL_WAITTIME;
```
Comment (thijskh) [inline judge/judgedaemon.main.php:582]: Same as above

**Review by thijskh — CHANGES_REQUESTED** [review]

File: judge/judgedaemon.main.php [hunk 7]
```diff
@@ -576,12 +579,14 @@ function fetch_executable(
...
         if (!$endpoint['waiting']) {
             $dosleep = false;
+            $waittime = INITIAL_WAITTIME_MS;
             break;
         }
     }
     // Sleep only if everything is "waiting" and only if we're looking at the first endpoint again
     if ($dosleep && $currentEndpoint==0) {
-        sleep($waittime);
+        usleep($waittime);
```
[the remark this answers is not in the document]
Comment (mvr320) [reply judge/judgedaemon.main.php:588]: I checked and you're indeed right.

**Comment by thijskh** [comment]
This is getting quite complex with all the multiplication factors. Also usleep might not really sleep and can still cause cpu activity.

How about this simpler approach:
- define sleeptime as 100 ms
- if  sleeptime < 1 s, sleep for sleeptime ms with usleep()
- else sleep() for the floor of sleeptime/1M maxed to 5 s
- multiply sleeptime by two and retry
- when getting work reset sleeptime to initial

**Review by meisterT — APPROVED** [review]

**Pull request merged**